dbms/src/Flash/tests/gtest_compute_server.cpp

```diff
@@ -849,6 +872,67 @@ try
 }
 CATCH
 
+TEST_F(ComputeServerRunner, testMinTSOActiveSetSoftLimit)
+try
+{
+    UInt64 active_set_soft_limit = 2;
+    context.context->setSetting("task_scheduler_active_set_soft_limit", active_set_soft_limit);
+    startServers(1);
+    setCancelTest();
+    ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 0);
+    std::vector<std::thread> running_queries;
+    std::vector<MPPQueryId> query_ids;
+    try
+    {
+        /// case 1, min tso can be added
+        for (size_t i = 0; i < active_set_soft_limit; i++)
+        {
+            addOneQuery(i + 10, running_queries, query_ids);
+        }
+        using namespace std::literals::chrono_literals;
+        std::this_thread::sleep_for(2s);
+        ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 2);
+        addOneQuery(1, running_queries, query_ids);
+        std::this_thread::sleep_for(2s);
+        ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 3);
+        for (const auto & query_id : query_ids)
+            MockComputeServerManager::instance().cancelQuery(query_id);
+        for (auto & t : running_queries)
+            t.join();
+        running_queries.clear();
+        query_ids.clear();
+        /// case 2, non-min tso can't be added
+        for (size_t i = 0; i < active_set_soft_limit; i++)
+        {
+            addOneQuery((i + 1) * 20, running_queries, query_ids);
+        }
+        using namespace std::literals::chrono_literals;
+        std::this_thread::sleep_for(2s);
+        ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 2);
+        addOneQuery(30, running_queries, query_ids);
+        std::this_thread::sleep_for(2s);
+        ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 2);
+        /// cancel 1 running query
+        MockComputeServerManager::instance().cancelQuery(query_ids[0]);
+        running_queries[0].join();
+        std::this_thread::sleep_for(2s);
+        ASSERT_TRUE(TiFlashMetrics::instance().tiflash_task_scheduler.get(tiflash_task_scheduler_metrics::type_active_queries_count).Value() == 2);
+        for (size_t i = 1; i < running_queries.size(); i++)
+            MockComputeServerManager::instance().cancelQuery(query_ids[i]);
+        for (size_t i = 1; i < running_queries.size(); i++)
+            running_queries[i].join();
+    }
+    catch (...)
+    {
+        for (const auto & query_id : query_ids)
+            MockComputeServerManager::instance().cancelQuery(query_id);
+        for (auto & t : running_queries)
+            if (t.joinable())
+                t.join();
```

Comment:
Is catch exception expected?

Reply:
Catch exception here is to avoid destruct of non-joined thread crash the process, I will rethrow the exception after all threads are joined.